ext/mbstring/mbstring.c

```diff
@@ -2282,6 +2289,256 @@ PHP_FUNCTION(mb_output_handler)
+	 */
+	if (mbfl_encoding->flag & MBFL_ENCTYPE_SBCS) { /* 1 byte */
+		mb_len = string.len;
+		size_t chunk_len = (size_t)split_length; /* chunk length in bytes */
+		if(chunk_len >= string.len){
+			chunk_len = string.len; /* chunk length can't be bigger then string */
+			chunks = 1;
+			array_init_size(return_value, chunks);
+			add_next_index_stringl(return_value, p, chunk_len);
+		} else{
+			chunks = mb_len / (size_t)split_length;
+			if(string.len % chunk_len == 0){
+				array_init_size(return_value, chunks);
+				for(size_t i = 0; i < chunks; p += chunk_len, ++i){
+					add_next_index_stringl(return_value, p, chunk_len);
+				}
+			} else{
+				array_init_size(return_value, chunks + 1);
+				for(size_t i = 0; i < chunks; p += chunk_len, ++i){ /* split cycle except last chunk */
+					add_next_index_stringl(return_value, p, chunk_len);
+				}
+				add_next_index_stringl(return_value, p, last - p); /* last chunk */
+			}
+		}
+	} else if (mbfl_encoding->flag & (MBFL_ENCTYPE_WCS2BE | MBFL_ENCTYPE_WCS2LE | MBFL_ENCTYPE_MWC2BE
+	| MBFL_ENCTYPE_MWC2LE)) { /* 2 bytes */
+		mb_len = string.len>>1; /* string length in chars */
+		size_t chunk_len = split_length<<1; /* chunk length in bytes */
+		if(chunk_len >= string.len){
+			chunk_len = string.len; /* chunk length can't be bigger then string */
+			chunks = 1;
+			array_init_size(return_value, chunks);
+			add_next_index_stringl(return_value, p, chunk_len);
+		} else{
+			chunks = mb_len / (size_t)split_length;
+			if(string.len % chunk_len == 0){
+				array_init_size(return_value, chunks);
+				for(size_t i = 0; i < chunks; p += chunk_len, ++i){
+					add_next_index_stringl(return_value, p, chunk_len);
+				}
+			} else{
+				array_init_size(return_value, chunks + 1);
+				for(size_t i = 0; i < chunks; p += chunk_len, ++i){ /* split cycle except last chunk */
+					add_next_index_stringl(return_value, p, chunk_len);
+				}
+				add_next_index_stringl(return_value, p, last - p); /* last chunk */
+			}
+		}
+	} else if (mbfl_encoding->flag & (MBFL_ENCTYPE_WCS4BE | MBFL_ENCTYPE_WCS4LE )) { /* 4 bytes */
+		mb_len = string.len>>2;
+		size_t chunk_len = split_length<<2; /* split length in bytes */
+		if(chunk_len >= string.len){
+			chunk_len = string.len; /* chunk length can't be bigger then string */
+			chunks = 1;
+			array_init_size(return_value, chunks);
+			add_next_index_stringl(return_value, p, chunk_len);
+		} else{
+			chunks = mb_len / (size_t)split_length;
+			if(string.len % chunk_len == 0){
+				array_init_size(return_value, chunks);
+				for(size_t i = 0; i < chunks; p += chunk_len, ++i){
+					add_next_index_stringl(return_value, p, chunk_len);
+				}
+			} else{
+				array_init_size(return_value, chunks + 1);
+				for(size_t i = 0; i < chunks; p += chunk_len, ++i){ /* split cycle except last chunk */
+					add_next_index_stringl(return_value, p, chunk_len);
+				}
+				add_next_index_stringl(return_value, p, last - p); /* last chunk */
+			}
+		}
```

Comment:
It should be possible to combine these three cases by using the flags to determine the size of a character (1, 2 or 4) and then using that value. I.e. mb_len = string.len / char_len and chunk_len = split_length * char_len, etc. This will cut down the code by a lot.

Reply:
Thank you. I have combined all fixed width encodings.